### PR TITLE
REST API: Change likes and sharing post meta to REST fields

### PR DESCRIPTION
**Pull request by codebykat**
This is an update to https://github.com/Automattic/jetpack/pull/11196, further groundwork to support migrating the Likes and Sharing metabox to Gutenberg. After working to implement this logic within Calypso, it became obvious that we should offload the complexity to the API instead.

See https://github.com/Automattic/wp-calypso/pull/29744 for the front-end block code and D23763-code for prior discussion on the WPCOM side.

#### Changes proposed in this Pull Request:
* Rather than using `register_meta`, we need to use `register_rest_field` so that we can add a response with custom callbacks for get and update
* Instead of simply passing along the existing (confusing) `post_meta` fields, this PR coerces them into booleans that make sense

#### Testing instructions:
* In both Gutenberg and the Classic Editor, verify that the old (backwards-compatible) Likes and Sharing metabox still displays, and works as intended.
* In Jetpack Sharing settings, enable the toggles to show Likes and Sharing Buttons.
* Request the post endpoint via the REST API. Verify that `jetpack_likes_enabled` and `jetpack_sharing_enabled` show under the meta keys:

screen shot 2019-02-07 at 12 48 47 am

* Experiment with different combinations of sitewide like and sharing settings and post settings, and ensure the values returned by the API reflect reality

#### Proposed changelog entry for your changes:

* REST API: Change likes and sharing post meta keys to jetpack_likes_enabled and jetpack_sharing_enabled keys

### CODE REVIEW

File: modules/likes.php [hunk 1]
```diff
@@ -557,24 +557,46 @@ function admin_bar_likes() {
 	}
 }
 
+/**
+ * Callback to get the value for the jetpack_likes_enabled field.
+ */
+function jetpack_post_likes_get_value( array $post ) {
+	$post_likes_switched = get_post_meta( $post['id'], 'switch_like_status', true );
+
+	$sitewide_likes_enabled = (bool) apply_filters( 'wpl_is_enabled_sitewide', ! get_option( 'disabled_likes' ) );
```
Comment (jeherve) [inline modules/likes.php]: Could you add this above the filter here and below, to avoid tripping our parser?
`/** This filter is documented in modules/likes.php */`

File: modules/sharedaddy/sharing.php [hunk 2]
```diff
@@ -547,25 +547,48 @@ public function management_page() {
 	}
 }
 
+/**
+ * Callback to get the value for the jetpack_sharing_enabled field.
+ *
+ */
+function jetpack_post_sharing_get_value( array $post ) {
```
Comment (jeherve) [inline modules/sharedaddy/sharing.php]: It may just be me, but I find those function names a bit confusing, as I don't know right away what value this is referring to. It becomes even more confusing with values like `sharing_disabled`, set to true for example, that we negate to say that `jetpack_sharing_enabled` is... false. 😌 

What do you think?

**Review by jeherve — CHANGES_REQUESTED** [review]
Just some minor comments. I also wanted to point out that since #10600, we also have `WPCOM_REST_API_V2_Field_Controller` as a wrapper for `register_rest_field`. Here is how it's used today:
https://github.com/Automattic/jetpack/tree/5323005738d3c4a5a672f1d30be6875f4fab483b/_inc/lib/core-api/wpcom-fields

**Comment by mmtr** [comment]
Code looks good to me, nice work @codebykat! 

However, I get an incorrect value for the `jetpack_likes_enabled` field under these circumstances:

- Sharing and Like buttons are enabled in Jetpack > Settings:
screen shot 2019-02-07 at 15 56 43

- Sharing and Likes are disabled in a specific post (I created this post after enabling the global settings):
screen shot 2019-02-07 at 15 59 23

When I request the post data via the API, `jetpack_likes_enabled` is `true`. I expected to see a `false` value since the "Show likes" checkbox is unticked. I got the same results using either Gutenberg or the Classic editor.

image


Note that I have D24043-code applied with the API sandboxed and I run Jetpack locally (with docker + ngrok).

**Comment by gwwar** [comment]
> I also wanted to point out that since #10600, we also have `WPCOM_REST_API_V2_Field_Controller` as a wrapper for `register_rest_field`

@jeherve is this a non-blocking comment?

**Comment by jeherve** [comment]
> is this a non-blocking comment?

Yup, just pointing it out for reference, and for our future selves, as it may be more convenient in some scenarios. :)

File: modules/sharedaddy/sharing.php (hunk 2 above)
Comment (codebykat) [reply modules/sharedaddy/sharing.php]: Well, that's the confusion I'm trying to eliminate by renaming the post_meta for the API keys. I'm open to suggestions on the function names, I just tried to name all the callbacks according to the same pattern.

**Comment by kwight** [comment]
I've tested by applying this branch to my local Docker+ngrok setup, and then queried using the developer console (seemed to work well!) to get a single test post (`GET wp/v2/sites/:siteId/posts/:postId`).

Changing the global settings result in the fields being present or not (eg. the setting being off meant the field would not be in the API result, as opposed to being present and `false` – I assume this is the desired behaviour?). Changes to individual posts worked as expected for Sharing, but not for Likes; I couldn't get `jetpack_likes_enabled: false`, it was always `true` (or not present). This sounds like the quirk you mentioned?

**Comment by codebykat** [comment]
Ah sounds like there might be a bug with Likes -- I wonder if the filter is wrong on Jetpack? Investigating.

File: modules/likes.php (hunk 1 above)
Comment (codebykat) [reply modules/likes.php]: Done. FWIW I don't see that it actually *is* documented (though I do see other comments saying it is).

File: modules/sharedaddy/sharing.php (hunk 2 above)
Comment (codebykat) [reply modules/sharedaddy/sharing.php]: To save you the context reloading, these options are stored in post_meta on the post that looks like this:

Sharing: sharing_disabled meta key is set only if the global site setting is enabled, and we want to suppress sharing buttons on that particular post. Otherwise not set. It's not possible to enable sharing buttons when they are disabled globally.

Likes: Stored as an override of the global setting for the posts that have been set to invert the global setting. (I know. Terrible.) So if Likes are ON globally, a particular post can have them disabled with the switch_like_status meta; conversely, if Likes are OFF globally, a particular post can still have them enabled.

**Comment by codebykat** [comment]
And yes on Jetpack disabling the setting will not load the module, so the API key won't be present. We'd only expect to see a value of `false` when the global setting is enabled but the per-post setting is toggled off.

File: modules/likes.php (hunk 1 above)
Comment (jeherve) [reply modules/likes.php]: Ouch, that's a good point. The filter got moved to `modules/likes/jetpack-likes-settings.php` a while ago, but the references never got updated.

File: modules/sharedaddy/sharing.php (hunk 2 above)
Comment (gwwar) [reply modules/sharedaddy/sharing.php]: This might be a good candidate to leave a note, saying that this is complicated, and link to public documentation on the behavior. Non-blocking of course @codebykat .

**Comment by gwwar** [comment]
This is really close @codebykat! I think I'm seeing the same behavior while testing that @mmtr saw in https://github.com/Automattic/jetpack/pull/11298#issuecomment-461457631.

Basically sharing appears to be reflected properly, but when toggling the likes property on a post it's never false for me, unless I modify the likes setting under sharing buttons in Calypso:
screen shot 2019-02-07 at 6 35 18 pm

**Comment by codebykat** [comment]
I FOUND IT and it's REALLY AWFUL.

On Jetpack the `switch_likes_setting` boolean meta is treated as a TERNARY option. So it is either unset (`get_post_meta` returns an empty string), `true`, or `false`. In the case where likes are disabled globally and the user wants to enable them on a particular post, the meta is set to **true**. Conversely, if likes are enabled globally and disabled on a particular post, the meta is set to **false**. If the meta is not set at all, we follow the global setting.

This is actually somewhat great in that it solves the problem of what happens when the global site setting gets changed. On Jetpack, an unset value means "No overrides -- follow the global setting." Whereas a set value allows us to determine what the user originally intended to happen, and proceed accordingly.

On the front-end, the check to display like buttons utilizes `empty()`, which is kind of genius as it resolves to `false` in the case of both `false` and unset.

On WPCOM, where we simply set the meta to `true` if the user wants to override the global setting, there is a bug when the global setting is changed.

However, it is of course *terrible* that this boolean currently means different things on different platforms.

I updated this PR with a POC solution that does different logic depending on whether this is a Jetpack or WPCOM environment, but tomorrow I plan to see if I can easily bring the WPCOM behavior in line with the Jetpack behavior and thus be able to simplify things again.

**Comment by codebykat** [comment]
After thinking about this a little, here's the issue on WPCOM that we'll need to either solve or accept that it will break for some users:

### Current status on WPCOM:

likes enabled globally, 
- a post with likes enabled => no meta
- a post with likes disabled => meta=1

likes disabled globally,
- a post with likes enabled => meta=1
- a post with likes disabled => no meta

### THEN: we change the behavior on WPCOM to follow Jetpack. So now:

likes enabled globally,
- a post with likes enabled => no meta => no change 👍 
- a post with likes disabled => meta=1 => likes are now enabled erroneously ❌

likes disabled globally,
- a post with likes enabled => meta=1 => likes are enabled, correctly ✅
- a post with likes disabled => no meta => no change 👍 


I also need to test more and verify whether this is different on wp-admin and Calypso; I suspect it's an API bug (as the metabox code is shared) and thus may be Calypso-only.

**Review by mmtr — DISMISSED** [review]
This is working well now. It wasn't trivial to figure out the differences between WP.com and Jetpack, good job @codebykat!

> I plan to see if I can easily bring the WPCOM behavior in line with the Jetpack behavior and thus be able to simplify things again.

That's great! But I don't consider that a blocker, so I'd say it is ok to simplify the behavior on a follow-up PR.

File: modules/sharedaddy/sharing.php (hunk 2 above)
Comment (jeherve) [reply modules/sharedaddy/sharing.php]: Thanks for the additional info. That's helpful. I am not quite sure how to make that easier, but a note may help, as Kerry suggested. 🙇

File: modules/likes.php [hunk 3]
```diff
@@ -557,24 +557,85 @@ function admin_bar_likes() {
...
+	/** This filter is documented in class.json-api-endpoints.php */
+	$is_jetpack = true === apply_filters( 'is_jetpack_site', false, get_current_blog_id() );
+
+	if ( $should_switch_status ) {
+		// WPCOM sites set the meta to 1 if the user is flipping the global setting
+		$meta_value = 1;
+		if ( $is_jetpack ) {
+			// set the meta to 0 if the user wants to disable likes, 1 if user wants to enable
+			$meta_value = $enable_post_likes;
+		}
```
Comment (dmsnell) [inline modules/likes.php]: this isn't a blocker, but in this pattern…

```php
$var = A;
if ( $condition ) {
	$var = B;
}
```

…I find it help to assign with one non-mutating binding and then we don't have to track reassignments in our head when reading the code, plus it's shorter and there's less visual noise.

```php
// Jetpack is a force-override while on WPCOM we indicate that we should flip
$meta_value = $is_jetpack ? $enable_post_likes : 1;
```

**Review by gwwar — DISMISSED** [review]
Awesome work @codebykat ! Great investigation on the different behaviors between wpcom and Jetpack. What a journey! The API response is now very clear to read too 💖 

I tested this with a variety of combinations of setting post options, global settings, and everything aligned on the published view and api response. 💯 

bmo

**Review by jeherve — DISMISSED** [review]
This seems to work well in my tests. Merge when ready! 🚢

File: modules/likes.php [hunk 4]
```diff
@@ -557,24 +557,85 @@ function admin_bar_likes() {
...
+ * Callback to get the value for the jetpack_likes_enabled field.
+ */
+function jetpack_post_likes_get_value( array $post ) {
+	$post_likes_switched = get_post_meta( $post['id'], 'switch_like_status', true );
+
+	/** This filter is documented in modules/likes.php */
+	$sitewide_likes_enabled = (bool) apply_filters( 'wpl_is_enabled_sitewide', ! get_option( 'disabled_likes' ) );
+
+	/** This filter is documented in class.json-api-endpoints.php */
+	$is_jetpack = true === apply_filters( 'is_jetpack_site', false, get_current_blog_id() );
```
Comment (dmsnell) [inline modules/likes.php]: is our use of `get_current_blog_id()` safe here? we ran into issues [in the VideoPress code](https://github.com/Automattic/jetpack/pull/11193/files#diff-db05a9485b8ce5e4945e32ff8cfd34b6R41) because for Jetpack sites it was returning values like `1` and `2` instead of the WordPress.com values returned from `get_option( 'ID' )`

File: modules/likes.php [hunk 5]
```diff
@@ -557,24 +557,85 @@ function admin_bar_likes() {
...
+ * Callback to set switch_like_status post_meta when jetpack_likes_enabled is updated.
+ */
+function jetpack_post_likes_update_value( $enable_post_likes, $post_object ) {
+	/** This filter is documented in modules/likes.php */
+	$sitewide_likes_enabled = (bool) apply_filters( 'wpl_is_enabled_sitewide', ! get_option( 'disabled_likes' ) );
+
+	$should_switch_status = $enable_post_likes xor $sitewide_likes_enabled;
+
+	/** This filter is documented in class.json-api-endpoints.php */
+	$is_jetpack = true === apply_filters( 'is_jetpack_site', false, get_current_blog_id() );
```
Comment (dmsnell) [inline modules/likes.php]: see note above on the use of `get_current_blog_id()`

**Comment by codebykat** [comment]
Follow-up sleuthing: this was changed in Jetpack here https://github.com/Automattic/jetpack/pull/9076 and it appears that a key fusioned file has diverged. I'm checking with the Jetpack folks about the best way to get this back in sync. If we can just sync this change back to WPCOM that will solve some problems.

File: modules/likes.php (hunk 4 above)
Comment (codebykat) [reply modules/likes.php]: Hmm I'm not sure 🤔 I copied it from elsewhere in Jetpack, but now that I look, that turns out to be from within the API. I'll follow up, but am hoping to eliminate the need for it anyway..

Comment (jeherve) [reply modules/likes.php]: As far as I can tell we only use the `is_jetpack_site` filter in one place in Jetpack, and we don't rely on the blog ID at all there.
https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/jetpack.php#L114

We do use it on WordPress.com though, but that should not be an issue there?

File: modules/likes.php (hunk 1 above)
Comment (codebykat) [reply modules/likes.php]: I was back in here anyway so I updated the references in this particular file.

File: modules/sharedaddy/sharing.php (hunk 2 above)
Comment (codebykat) [reply modules/sharedaddy/sharing.php]: I added some more verbose comments to the function definitions.

**Comment by codebykat** [comment]
I've updated this to remove the platform conditionals. It now depends on landing https://href.li/?https://github.com/Automattic/jetpack/pull/11336 and https://href.li/?https://github.com/Automattic/jetpack/pull/11346 (and their WPCOM counterparts).

**Comment by gwwar** [comment]
Writing this out so I don't confuse myself:

WPCOM Current Behavior, (Global xor Meta)

| Global  | Meta | Show Likes |
| ------------- | ------------- |------------- |
| True  | Empty/False  | True |
| True  | True  | False |
| False  | Empty/False  | False|
| False  | True  | True |

Jetpack Current Behavior, Use meta value, otherwise fallback to global.

| Global  | Meta | Show Likes |
| ------------- | ------------- |------------- |
| True  | Empty | True |
| True  | True  | True |
| True  | False  | False |
| False  | Empty | False |
| False  | True | True |
| False  | False  | False |

**Comment by codebykat** [comment]
I've fixed the tests in D24043-code, but not sure how to get the bot to pick that up. Anyone know?

**Comment by mdawaffe** [comment]
> I've fixed the tests in D24043-code, but not sure how to get the bot to pick that up. Anyone know?

@dereksmart, if we run it through fusion again, will it overwrite the differential, or just run the tests?

@codebykat, whatever Derek's answer, we can fix it - I'm just not sure which of two ways is the right way :)

**Comment by mmtr** [comment]
I have rerun fusion thinking it will just run the tests and indeed has overwritten the differential again :( 

Sorry, @codebykat ! I added the tests back so you don't need to spend time re-fixing them but I needed to commandeer the diff, so feel free to take control of the revision and become its author again.

**Comment by codebykat** [comment]
Re-Fusioning didn't work, so I guess we need whatever the other way is 😛

**Comment by mdawaffe** [comment]
Other way accomplished! :)

**Comment by kwight** [comment]
I've given this another quick runthrough, after merging #11346 and doing a local rebase. Different combinations of Likes and Shares in both Gutenberg and Classic all reflected properly in both admin UI and the front-end 🎉 

One thing I noticed, was that the response to the Gutenberg `POST` from clicking Update (eg. `/wp-json/wp/v2/posts/:postID` returns the previous settings (if I've activated Likes, `jetpack_likes_enabled` will be `false`), but the UI and front-end will be correct.

**Comment by jeherve** [comment]
> One thing I noticed, was that the response to the Gutenberg `POST` from clicking Update (eg. `/wp-json/wp/v2/posts/:postID` returns the previous settings (if I've activated Likes, `jetpack_likes_enabled` will be `false`), but the UI and front-end will be correct.

Would #11387 help with this?

**Comment by codebykat** [comment]
> Would #11387 help with this?

It shouldn't matter, no. 🤔 

I deployed a bunch of the infrastructure changes today, I will re-test this tomorrow and see if I can duplicate the behavior @kwight was seeing or if it was due to dependencies being in an inconsistent state.

**Comment by mmtr** [comment]
> One thing I noticed, was that the response to the Gutenberg `POST` from clicking Update (eg. `/wp-json/wp/v2/posts/:postID` returns the previous settings (if I've activated Likes, `jetpack_likes_enabled` will be `false`), but the UI and front-end will be correct.

I can reproduce this:

feb-25-2019 14-17-55

Maybe this can be solved if we return a value on the update callbacks?

File: modules/likes.php [hunk 6]
```diff
@@ -557,24 +557,82 @@ function admin_bar_likes() {
...
+ * When it is set to 1, we enable likes on the post, regardless of the global setting.
+ */
+function jetpack_post_likes_update_value( $enable_post_likes, $post_object ) {
+	/** This filter is documented in modules/jetpack-likes-settings.php */
+	$sitewide_likes_enabled = (bool) apply_filters( 'wpl_is_enabled_sitewide', ! get_option( 'disabled_likes' ) );
+
+	$should_switch_status = $enable_post_likes xor $sitewide_likes_enabled;
+
+	/** This filter is documented in class.json-api-endpoints.php */
+	$is_jetpack = true === apply_filters( 'is_jetpack_site', false, get_current_blog_id() );
```
Comment (mmtr) [inline modules/likes.php]: This is no longer needed, right?

**Comment by codebykat** [comment]
BUG FOUND AND FIXED. I also added explicit returns and schema after looking at some of the example code for the API.

This is ready for review again 🤞

**Comment by codebykat** [comment]
Also noting this patch doesn't apply cleanly on WPCOM because I accidentally reverted the `register_meta` addition on the WPCOM side. Which doesn't actually matter since this code removes that call anyway, but I'm not sure what the best thing is to do wrt the automated tests.

**Comment by mmtr** [comment]
> BUG FOUND AND FIXED. I also added explicit returns and schema after looking at some of the example code for the API.

Thanks @codebykat! However I still see the weird behavior with the `POST` requests (and only with the `POST` requests). Code looks good to me, and the `GET` requests return the correct values so I don't consider this issue a blocker.

Actually, I'm starting to think that it might be a core issue caused by making a `POST` request without updating the new fields.

> Also noting this patch doesn't apply cleanly on WPCOM because I accidentally reverted the `register_meta` addition on the WPCOM side. Which doesn't actually matter since this code removes that call anyway, but I'm not sure what the best thing is to do wrt the automated tests.

I think you can go ahead and manually update the diff so the changes are synced.

File: modules/sharedaddy/sharing.php [hunk 7]
```diff
@@ -552,25 +552,55 @@ public function management_page() {
...
+ *
+ * When the sharing_disabled post_meta is unset, we follow the global setting in Sharing.
+ * When it is set to 1, we disable sharing on the post, regardless of the global setting.
+ * It is not possible to enable sharing on a post if it is disabled globally.
+ *
+ */
+function jetpack_post_sharing_update_value( $enable_sharing, $post_object ) {
+	if ( $enable_sharing ) {
+		// delete the override if we want to enable sharing
+		delete_post_meta( $post_object->ID, 'sharing_disabled' );
```
Comment (mmtr) [inline modules/sharedaddy/sharing.php]: Should we return a value here as we do in `jetpack_post_likes_update_value`?

File: modules/sharedaddy/sharing.php [hunk 8]
```diff
@@ -552,25 +552,55 @@ public function management_page() {
...
-		'post', 'sharing_disabled',
+function jetpack_post_sharing_register_rest_field() {
+	register_rest_field(
+		'post', 'jetpack_sharing_enabled',
 		array(
-			'type'			=> 'boolean',
-			'single'		=> true,
-			'show_in_rest'	=> true,
+			'get_callback' => 'jetpack_post_sharing_get_value',
+			'update_callback' => 'jetpack_post_sharing_update_value',
```
Comment (mmtr) [inline modules/sharedaddy/sharing.php]: I think it'd be good to also define the schema here.

Comment (codebykat) [reply modules/sharedaddy/sharing.php]: Ah right I did this on the WPCOM side and then didn't sync this file over. Updated.

**Comment by codebykat** [comment]
@mmtr Hmm. I tested this from the API console, and my fixes from last night did make the correct values show up from both POST and GET requests. Can you share the specific reproduction steps and whether you have Like buttons enabled or disabled globally? Also, are you seeing this with the sharing value or just likes?

**Comment by codebykat** [comment]
> I think you can go ahead and manually update the diff so the changes are synced.

Yeah, I did that, but it's showing up in the "Some checks were not successful" box because it couldn't sync it over. I don't know whether or not that blocks merging.

**Comment by mmtr** [comment]
> @mmtr Hmm. I tested this from the API console, and my fixes from last night did make the correct values show up from both POST and GET requests. Can you share the specific reproduction steps and whether you have Like buttons enabled or disabled globally? Also, are you seeing this with the sharing value or just likes?

I get the results while testing on a Jetpack site and updating a post from the UI as you can see on the GIF I shared above:

feb-25-2019 14-24-48

- Go to Posts > Add new
- Enter some content and publish the post or save it as draft
- Toggle any of the Likes/Shares checkboxes
- Change some content in the post and update it
- Open the browser devtools and go to the Network tab
- Check the response of the `POST` request to the API that is triggered when we update the post

**Comment by kwight** [comment]
Hm, what actually updates the values in the db? I can't tell; in fact, I can't get the code in `jetpack_post_likes_update_value` to execute at all. And I think @mmtr and I are seeing what we're seeing because we're both looking at `POST` calls to `:domain/wp-json/wp/v2/posts/272?_locale=user`, which doesn't seem to trigger the update at all, it just does a read of those values via `jetpack_post_likes_get_value` (likely before the actual update mechanism does its thing).

>  I tested this from the API console, and my fixes from last night did make the correct values show up from both POST and GET requests.

@kat what is the API call you're making to see the correct results? Is it the same as the `POST` one above? I wonder if the confusion is that we're all looking at different things 🙃

**Comment by codebykat** [comment]
On a sandboxed WPCOM site, `POST wp/v2/sites/sandboxed.wpcom.site/posts/post_ID` with the payload `jetpack_likes_enabled` set to 0 or 1.

Does this happen for y'all on WPCOM sites or just on Jetpack? The logic should be the same 🤔 

I think you're right @kwight that the update does call `GET` to retrieve the new value, but only after running the update code. Is there possibly a missing piece in having it sandboxed on the WPCOM API side as well? I realized I've been testing with the patch applied in both places.

**Comment by codebykat** [comment]
Also can you duplicate the issue with the sharing setting, or is it just Likes?

**Comment by kwight** [comment]
Ah OK, I think we've been testing two different ways.

@mmtr and I have been changing values in the UI, then hitting Update; this fires the `POST` call, but the Likes and Shares values are not updated by this call, they're updated **after** it in the `meta-box-loader` call which does not go through the API (it's a `POST` to `wp-admin/post.php`). This is why the return values we're seeing are out-of-date, and I can't get code in the update part of this PR to fire.

@kat you've been making a direct API call with the fields themselves directly in the body of the `POST`, meaning they _are_ handled by this PR's `jetpack_post_likes_update_value` function. Done this way, the return value _is_ correct, as it should be.

TL;DR, this PR is good to go, as far as I'm concerned :)

> Does this happen for y'all on WPCOM sites or just on Jetpack?

I've just been testing this PR against a local ngrokked JP site, so I haven't had the API or anything sandboxed. From what I can tell, these calls hit the site directly, and don't have anything to do with `public-api.wordpress.com`.

**Review by kwight — APPROVED** [review]

**Review by mmtr — APPROVED** [review]
Thanks for the analysis @kwight ! All looking good then :)

Awesome work @codebykat!

**Comment by kwight** [comment]
@jeherve Anything else we need to merge this, beyond one of you to approve? We've forced the `wpcom` check to pass, because syncing needed manual intervention dotcom-side.

**Comment by kraftbj** [comment]
With everyone else approving, let me give it a quick last look.

**Review by kraftbj — APPROVED** [review]
I tried various combinations of status changes using Classic Editor and the Jetpack console. Meta via the REST endpoint looked as expected per testing call.

**Comment by kraftbj** [comment]
@codebykat @kwight Please merge this at will when y'all are ready.

**Comment by gwwar** [comment]
Excellent shepherding and bugfixing @codebykat ! 🎉 

💖 from the great code reviews here too!
